torchtune/models/flamingo/_encoders.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+from torch import nn
+
+from torchtune.modules import (
+    Fp32LayerNorm,
+    GroupedQueryAttention,
+    TanhGate,
+    TransformerSelfAttentionLayer,
+)
+
+
+class FlamingoVisionAdapter(nn.Module):
+    def __init__(
+        self,
+        embed_dim: int,
+        num_layers: int,
+        num_heads: int,
+        proj_in: int,
+        proj_out: int,
+    ) -> None:
+        super().__init__()
+
+        mlp_ratio = 4
+        hidden_dim = int(mlp_ratio * embed_dim)
+        head_dim = embed_dim // num_heads
+        num_kv_heads = num_heads
+
+        transformer_layers = []
+        for idx in range(1, num_layers + 1):
+            self_attn = GroupedQueryAttention(
+                embed_dim=embed_dim,
+                num_heads=num_heads,
+                num_kv_heads=num_heads,
+                head_dim=head_dim,
+                q_proj=nn.Linear(embed_dim, num_heads * head_dim, bias=True),
+                k_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=True),
+                v_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=True),
+                output_proj=nn.Linear(embed_dim, embed_dim, bias=True),
+                pos_embeddings=None,
+                attn_dropout=0.0,
+                default_causal_mask=False,
+            )
+
+            mlp = FlamingoMLP(
+                in_dim=embed_dim,
+                hidden_dim=hidden_dim,
+                out_dim=embed_dim,
+                act_layer=nn.GELU,
+            )
+
+            layer = TransformerSelfAttentionLayer(
+                attn=attn,
+                mlp=mlp,
+                attn_norm=Fp32LayerNorm(embed_dim, eps=1e-5),
+                mlp_norm=Fp32LayerNorm(embed_dim, eps=1e-5),
+                attn_scale=TanhGate(),
+                mlp_scale=TanhGate(),
+            )
+
+            transformer_layers.append(layer)
+
+        self.projection = nn.Linear(proj_in, proj_out)
+
+    def forward(self, x: torch.Tensor, hidden_states: torch.Tensor) -> torch.Tensor:
+
+        bsz, n_ims, n_tiles, n_tokens, embed_dim = x
```

Comment:
Should be this?
```suggestion
        bsz, n_ims, n_tiles, n_tokens, embed_dim = x.shape
```

Reply:
Yeh it would be good to add some quick tests to make sure these are working as expected